component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/SchemaImpl.java

```diff
@@ -39,6 +41,19 @@
 
     private List<Entry> entries;
 
+    private Map<String, String> props = new LinkedHashMap<>(0);
+
+    public SchemaImpl(final Type type, final Schema schema, final List<Entry> entries) {
+        this.type = type;
+        elementSchema = schema;
+        this.entries = entries;
+    }
```

Comment:
Like this, with "@AllArgsConstructor" and "@Data" annotations, SchemaImpl instance become mutable.
```java
new SchemaImpl(x, y, z, myMap); myMap.put(xx, zz);
schema.getProps().put(xx, zz)
```
1 . put explicit constructor : 
```java
public SchemaImpl(final Type type, final Schema schema, final List<Entry> entries, Map<String, String> xx) {
   Map<String, String> val = new ...;
   val.addAll(xx);
   this.props = Collections.unmodifiableMap(val);  // while we don't use [vavr lib](https://docs.vavr.io/) to have real immutable collection.
}
```
2. make props final field

Reply:
Initially, the schema is mutable. We don't have to make it immutable, this may break some stuff.

Reply:
You mean some code use "schema.getEntries().add( xxx );" instead of using Schema.Builder class ? (except that, Schema interface leaves no place to mutation)
On SchemaImpl : type, elementSchema and entries are not final, but it should.

Reply:
You're right but this is not in scope of the changes. Please create a new issue for that.

Reply:
done : https://jira.talendforge.org/browse/TCOMP-1876